src/test/java/com/squareup/javapoet/JavaFileTest.java

```diff
@@ -228,6 +228,27 @@
         + "}\n");
   }
 
+  @Test public void nestedClassAndSuperclassShareName() throws Exception {
+    String source = JavaFile.builder("com.squareup.tacos",
+        TypeSpec.classBuilder("Taco")
+            .superclass(ClassName.get("com.squareup.wire", "Message"))
+            .addType(TypeSpec.classBuilder("Builder")
+                .superclass(ClassName.get("com.squareup.wire", "Message", "Builder"))
+                .build())
+            .build())
+        .build()
+        .toString();
+    assertThat(source).isEqualTo(""
+        + "package com.squareup.tacos;\n"
+        + "\n"
+        + "import com.squareup.wire.Message;\n"
+        + "\n"
+        + "class Taco extends Message {\n"
+        + "  class Builder extends Message.Builder {\n"
```

Comment:
Particularly noteworthy: with this change it's now pretty trivial to turn on and off nested-class imports. So we could import `com.squareup.wire.Message.Builder` directly; whereas we couldn't as easily before.

The _catch_ with that is that if we turn it on, it's not very smart, and so we always get the deepest types. I think the current, top-level-policy mostly works.